be1-go/channel/federation/federation.go

```diff
@@ -475,6 +525,97 @@ func (c *Channel) processChallengeRequest(msg message.Message,
 	return nil
 }
 
+func (c *Channel) processFederationResult(msg message.Message,
+	msgData interface{}, s socket.Socket) error {
+	_, ok := msgData.(*messagedata.FederationResult)
+	if !ok {
+		return xerrors.Errorf("message %v is not a federation#result message",
+			msgData)
+	}
+
+	/*
+	   This state should probably be set after sending the FederationInit message?
+	*/
+	//if c.state != WaitResult {
+	//	return answer.NewInternalServerError(invalidStateError, c.state, msg)
+	//}
+
+	var federationResult messagedata.FederationResult
+
+	err := msg.UnmarshalData(&federationResult)
+	if err != nil {
+		return xerrors.Errorf("failed to unmarshal FederationResult data: %v", err)
+	}
+
+	if federationResult.Status != "success" {
+		if len(federationResult.Reason) > 0 {
+			return xerrors.Errorf("failed to establish federated connection: %v", federationResult.Reason)
+		}
+		return xerrors.Errorf("failed to establish federated connection")
+	}
+
+	var federationChallenge messagedata.FederationChallenge
+	err = federationResult.ChallengeMsg.UnmarshalData(&federationChallenge)
+	if err != nil {
+		return xerrors.Errorf("failed to unmarshal challenge from FederationResult data: %v", err)
+	}
+
+	challengeDataBytes, err := base64.URLEncoding.DecodeString(federationResult.ChallengeMsg.Data)
+	if err != nil {
+		return xerrors.Errorf("failed to decode challenge data in FederationResult: %v", err)
+
+	}
+
+	challengeSignatureBytes, err := base64.URLEncoding.DecodeString(federationResult.ChallengeMsg.Signature)
+	if err != nil {
+		return xerrors.Errorf("failed to decode challenge signature in FederationResult: %v", err)
+
+	}
+
+	remoteOrg := c.getRemoteOrganization(federationChallenge.Value)
+	remoteOrg.Lock()
+	defer remoteOrg.Unlock()
+	pkBytes, err := base64.URLEncoding.DecodeString(remoteOrg.organizerPk)
```

Comment:
We should check that the remoteOrg.state is WaitResult before trying to process the message